streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBWindowStore.java

```diff
@@ -48,11 +49,12 @@
 
     private static final long USE_CURRENT_TIMESTAMP = -1L;
 
-    private static class Segment extends RocksDBStore<byte[], byte[]> {
+    // use the ByteBuffer wrapper for underlying rocksDB keys since they are used for hashing data structures
+    private static class Segment extends RocksDBStore<ByteBuffer, byte[]> {
```

Comment:
I agree: rather than using mutable and error-prone ByteBuffers, I'd prefer to see a simple custom-built immutable type here. This could also permit the added benefit of caching their hashCodes (ByteBuffer doesn't do this), and perhaps even interning them to prevent retaining duplicate copies of identical values.